tests/backend/test_bosonic_backend.py

```diff
@@ -0,0 +1,566 @@
+        assert state.weights().shape == (num_weights,)
+        assert np.allclose(sum(state.weights()), 1)
+        assert state.means().shape == (num_weights, 2)
+        assert state.covs().shape == (num_weights, 2, 2)
+
+        # Weights, means and covs should be real
+        assert np.allclose(state.weights().real, state.weights())
+        assert np.allclose(state.means().real, state.means())
+        assert np.allclose(state.covs().real, state.covs())
+
+        # Covariance should be diagonal with these entries
+        cov_val = (1 - np.exp(-2 * eps)) / (1 + np.exp(-2 * eps)) * sf.hbar / 2
+        covs_compare = np.tile(cov_val * np.eye(2), (num_weights, 1, 1))
+        assert np.allclose(state.covs(), covs_compare)
+
+        # Means should be integer multiples of sqrt(pi*hbar)/2 times a damping
+        damping = 2 * np.exp(-eps) / (1 + np.exp(-2 * eps))
+        mean_ints = state.means() / (damping * np.sqrt(np.pi * sf.hbar) / 2)
+        # Round to make sure numerical errors are washed out
+        mean_ints = np.round(np.real_if_close(mean_ints), 10)
+        assert np.allclose(mean_ints % 1, np.zeros(mean_ints.shape))
+
+    @pytest.mark.parametrize("eps", EPS_VALS)
+    def test_gkp_logical(self, eps):
+        r"""Checks that logically equivalent GKP states have
+        Wigner functions that agree."""
+        x = np.linspace(-3 * np.sqrt(np.pi), 3 * np.sqrt(np.pi), 40)
+        p = np.linspace(-3 * np.sqrt(np.pi), 3 * np.sqrt(np.pi), 40)
+
+        # Prepare GKP 0 and apply Hadamard
+        prog_0H = sf.Program(1)
+        with prog_0H.context as q0:
+            sf.ops.GKP(epsilon=eps) | q0[0]
+            sf.ops.Rgate(np.pi / 2) | q0[0]
+
+        backend_0H = bosonic.BosonicBackend()
+        backend_0H.run_prog(prog_0H)
+        state_0H = backend_0H.state()
+        wigner_0H = state_0H.wigner(0, x, p)
+
+        # Prepare GKP +
+        prog_plus = sf.Program(1)
+        with prog_plus.context as qp:
+            sf.ops.GKP(state=[np.pi / 2, 0], epsilon=eps) | qp[0]
+
+        backend_plus = bosonic.BosonicBackend()
+        backend_plus.run_prog(prog_plus)
+        state_plus = backend_plus.state()
+        wigner_plus = state_plus.wigner(0, x, p)
+
+        assert np.allclose(wigner_0H, wigner_plus)
+
+    @pytest.mark.parametrize("eps", EPS_VALS)
+    def test_gkp_state_parity(self, eps):
+        r"""Checks that GKP states yield the right parity."""
+        prog = sf.Program(1)
+        with prog.context as q:
+            sf.ops.GKP(epsilon=eps) | q[0]
+
+        backend = bosonic.BosonicBackend()
+        backend.run_prog(prog)
+        state = backend.state()
+        assert np.allclose(state.parity_expectation([0]), 1, atol=r_fock)
+
+    def test_gkp_complex(self):
+        r"""Checks that trying to call a complex representation
+        raises an error.
+
+        This test can be updated once the complex representation is
+        implemented."""
+        with pytest.raises(NotImplementedError):
```

Comment:
Let's shrink the scope of this context. Right now any statement within in could raise a `NotImplementedError` error (e.g., even `prog = sf.Program(1)`).

Reply:
Done.